rfcs/20200812-multidim-matmul/README.md

```diff
@@ -0,0 +1,188 @@
+> (src, A), (weights, B), (dst, C) and (Tensor, matrix).
+
+## Motivation
+Frameworks request for NumPy-like matmul that can support multiplication of two
+Tensors with dimensions of tensors greater or equal to two. Currently, the
+existing MatMul primitive supports at most one batch dim (in other words 3D
+tensor).
+
+The proposed RFC is an attempt to support following additional semantics over
+the existing dnnl_matmul primitive:
+  - (i) Multiple batch dimensions and
+  - (ii) Broadcasting.
+
+In general operation can be described as:
+
+`C = A * B + bias` , where `*` denotes matrix-matrix multiplication of two inner
+dimensions.
+
+An example:
+```python
+import numpy as np
+
+a = np.random.rand(4, 1, 5, 6, 6)
+print(a.shape)
+# Out: (4, 1, 5, 6, 6)
+
+b = np.random.rand(3, 1, 2, 1, 6, 6)
+print(b.shape)
+# Out: (3, 1, 2, 1, 6, 6)
+
+print(np.matmul(a, b).shape) # A will be auto expanded to shape (1, 4, 1, 5, 6, 6).
+# Out: (3, 4, 2, 5, 6, 6)
+```
+For general Broadcasting rules, refer to this [excellent article](https://numpy.org/doc/1.19/user/basics.broadcasting.html).
+
+> Note: GPU or other arch support is out of scope for this RFC.
+
+## Options
+
+Option 0.x are existing known workarounds to support this feature (sometimes
+partially).
+
+### Option 0.a Use MKL
+Currently TensorFlow uses `cblas_?gemm_batch` [API](https://software.intel.com/content/www/us/en/develop/articles/introducing-batch-gemm-operations.html) from MKL binary with array of
+pointers to matrices for A and B.
+
+- Pros:
+  - Uses existing optimized MKL implementation.
+  - Optimal performance.
+  - No maintenance/testing burden on oneDNN.
+- Cons:
+  - Dependency on MKL binary.
+
+### Option 0.b Call GeMM in a loop
+Example: A = {3 x M x K} and B = {1 x K x N}
+```python
+for i in range(3) :
+  C[i] = GeMM(A[i], B)
+```
+
+- Pros:
+  - Uses existing GeMM impl.
+  - No maintenance/testing burden on oneDNN.
+- Cons: Performance may not be optimal.
+
+### Option 1 Extend MatMul primitive
+
+This primitive is similar to oneMKL [API](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/blas-and-sparse-blas-routines/blas-like-extensions/cblas-gemm-batch-strided.html)
+`cblas_?gemm_batch_strided`, as all matrices have fixed stride.
+
+**Broadcasting**
+  - Support implicit broadcasting, but keep the restriction
+  `ndims(A) == ndims(B) == ndims(C) == ndims(bias)` (Note: Unfortunately, this
+  means that the example at the beginning of this doc is not directly supported.
+  The tensor A must be manually expanded by user to match number of dims of B).
+  - Broadcasting on dims `M`, `N` and `K` is invalid and not supported.
+
+**Layouts**
+  - Support only plain and transposed layouts. For example, in a 5D tensor only
+  `abcde` and `abced` layouts are supported. Requires adding two tags (one for
+  plain and one for transpose) for each dim if not already present to the
+  library.
```

Comment:
nit: strictly speaking we have `memory_desc_init_by_strides()`, so we can actually avoid introducing any new memory formats...

Reply:
(though this might be required for testing)

Reply:
I think, I also need it for tag checking to determine gemm_impl vs ref_impl deployment. Is there way to achieve this too without adding tags? If yes, would you prefer that instead of adding tags to library?

Reply:
> Is there way to achieve this too without adding tags?

Doesn't strides comparison work (at least in theory)?

> If yes, would you prefer that instead of adding tags to library?

It is a matter of usability / productivity. If tags are more efficient -- then I see no problem adding them.
Also, I guess benchdnn will need tags to do testing.

Sorry for the noise. Resolved.